torchrl/envs/transforms/transforms.py

```diff
@@ -3607,3 +3607,56 @@ def transform_input_spec(self, input_spec: CompositeSpec) -> CompositeSpec:
             if not self.create_copy:
                 del input_spec[in_key]
         return input_spec
+
+
+class Reward2GoTransform(Transform):
+    """Calculates the reward to go based on the episode reward and a discount factor.
+
+    As the Reward2GoTransform is only an inverse transform the in_keys will be directly used for the in_keys_inv.
+    The reward-to-go can be only calculated once the episode is finished. Therefore, the transform should be applied to the replay buffer
+    and not to the collector.
+
+    Args:
+        in_keys (list of str/tuples of str): the entries to rename
+        gamma (float or torch.Tensor): the discount factor
+
+    """
+
+    def __init__(
+        self,
+        gamma: Optional[Union[float, torch.Tensor]] = 1.0,
+        in_keys: Optional[Sequence[str]] = None,
+    ):
+        if in_keys is None:
+            in_keys = ["reward"]
+        out_keys = ["reward_to_go"]
+        super().__init__(
+            in_keys=in_keys,
+            in_keys_inv=in_keys,
+            out_keys_inv=out_keys,
+        )
+
+        if not isinstance(gamma, torch.Tensor):
+            gamma = torch.tensor(gamma)
+
+        self.register_buffer("gamma", gamma)
+
+    def _inv_call(self, tensordict: TensorDictBase) -> TensorDictBase:
+        episode_ends = torch.where(tensordict.get(("next", "done")))[0]
+        assert episode_ends.shape[0] > 0, "No episode ends found"
```

Comment:
We should not have `assert` in the code base. Here, I would use a `RuntimeError`.

Reply:
See my comment above for an implementation that checks that `all` trajectories have a `done` or `truncated` state.
Note: should they all *end* with a truncated or done? In other words, what happens if you cat two trajectories one after another and the second does not end?

Reply:
That's an open problem. The situation that we can have trajectories that didn't end yet are currently handled as if those ended. But a bigger problem is for a batch that consists of one whole trajectory that didn't end.  For example, the episode length is 2000 but we collect only 500 frames per batch. Which means there will be no True done state. For now, I managed this case with the RuntimeError and wrote explicitly that the frames_per_batch should at least be equal to the episode length. Once properly implemented with the Decision Transformer I will have to test what happens with episodes that didn't end properly (got cut off and didn't end with a done) but I think for most of the environments this will not be a huge problem like gym or mujoco.